Add MTU support across all platforms

robbym opened the pull request:
This PR will include a method to retrieve the negotiated MTU size on each platform.

Info grabbed from: [deviceplug/btleplug#246](https://github.com/deviceplug/btleplug/issues/246)

Windows - [GattSession.MaxPduSize](https://learn.microsoft.com/en-us/uwp/api/windows.devices.bluetooth.genericattributeprofile.gattsession.maxpdusize?view=winrt-22621)
Negotiation happens on user's behalf, and a GattSession.MaxPduSizeChanged event is emitted when MTU is updated.

macOS / iOS - [CBPeripheral.maximumWriteValueLength](https://developer.apple.com/documentation/corebluetooth/cbperipheral/1620312-maximumwritevaluelength)
Negotiation happens on user's behalf, have used peripheral.maximumWriteValueLength(for: .withoutResponse) reliably in the past. .withResponse has some issues. [See here.](https://stackoverflow.com/a/68389331)

Linux
Instead of a dedicated MTU property, BlueZ has decided to adopt the newer Bluetooth 5.2 standard. Example here: [Link](https://github.com/OpenBluetoothToolbox/SimpleBLE/blob/614a5af35cdd22ba08318cb776795ecd2da8c389/simpleble/src/backends/linux/PeripheralBase.cpp#L57)

Android - [BluetoothGatt.requestMtu](https://developer.android.com/reference/android/bluetooth/BluetoothGatt#requestMtu(int))
The way I understand it and have used it in the past, you call requestMtu with the largest you can support (e.g. 512), and it will negotiate the min(host.maxMtu, periph.maxMtu) and call onMtuChanged with the new mtu size.

TODO:

- [x] Windows
- [x] Linux
- [ ] Android
- [ ] macOS/iOS

robbym committed:
Implemented Windows MTU

qwandor commented on src/winrtble/peripheral.rs:
```diff
@@ -93,6 +94,7 @@ impl Peripheral {
                 adapter,
                 device: tokio::sync::Mutex::new(None),
                 address,
+                mtu: AtomicU16::new(23),
```
Where does this 23 come from?

robbym replied:
It's the default size of the MTU prior to negotiation. It includes the L2CAP header (4 bytes), so usable size would be 19.

robbym replied:
I think I'll add an initial assignment from MaxPduSize prior to the event handler registration.

qwandor replied:
Please add a constant for it, with a comment explaining this.

robbym committed:
Added initial MTU assignment prior to event handler registration

robbym committed:
Added a constant for default MTU size

laxian commented:
@robbym Hi, have you finish it on bluez?

robbym committed:
Moved constant to somewhere more accessible

robbym committed:
Bumped bluez-async version to bring in mtu property

robbym committed:
Implemented bluez version of mtu method

robbym commented:
@laxian I just pushed a working bluez implementation. I tested it on one of my BLE devices and it works, but let me know if there is any issues.

@qwandor I added the constant and now moved it to the api namespace so that each implementation can use it. Not really happy with where it currently lives. Maybe move it to common?

qwandor commented on src/bluez/peripheral.rs:
```diff
@@ -138,6 +138,17 @@ impl api::Peripheral for Peripheral {
+        let services = self.services.lock().unwrap();
+        for (_, service) in services.iter() {
+            for (_, characteristic) in service.characteristics.iter() {
+                return characteristic.info.mtu.unwrap();
```
Can this be different for different characteristics?

robbym replied:
In Bluetooth 5.2 and forward, yes. The reason they implemented it as a per-characteristic MTU is because of the enhanced ATT (EATT) bearer feature.

They mention it here: https://github.com/bluez/bluez/issues/199
I mention it here: https://github.com/deviceplug/btleplug/issues/246#issuecomment-1545186841

The majority of devices on the market are going to be using the unenhanced version (single MTU per connection), so I don't understand why bluez didn't add support for that first.

In any case, the implementation was based on how SimpleBle did it: https://github.com/OpenBluetoothToolbox/SimpleBLE/blob/614a5af35cdd22ba08318cb776795ecd2da8c389/simpleble/src/backends/linux/PeripheralBase.cpp#L57

qwandor replied:
Okay, seems reasonable given the API limitations.

qdot commented:
Windows parts of this are looking good so far. Would like to get all platforms in before we land it (otherwise I *will* have people on unsupported platforms yelling) but I'm fine to also do a point release whenever we get all the parts in.

qdot commented:
Converted this to a draft until we get all the platforms in. Just makes my life easier for the little amount of time I actually get to do triage on this project.

qdot commented:
One more quick update here: I brought us up to bluez-async 0.8.0 in v0.11.7, so mtu() is available now and we can drop that change here (it'll probably just skip the commit on rebase).

qwandor reviewed (approved):
BlueZ implementation seems fine.

qdot commented:
Ok, having to rebase this elsewhere while trying to get the other platforms added in, so:

- Closing this
- Rebasing and fixing conflicts in original commits
- Working out of `pr-407-mtu` to try to bring in all platforms
- Doing a followup to actually negotiate MTUs on all platforms.
